app/models/concerns/custom_field_addable.rb

```diff
@@ -0,0 +1,163 @@
+# JSON field on the model instance.
+# Assumes that a JSON db column named 'custom_data' exists.  (future: allow this column name to be customized)
+#
+
+
+module CustomFieldAddable
+  extend ActiveSupport::Concern
+
+  # todo: confirm if there are any negative consequences of potential duplicate includes of the Translatable module
+  include Translatable
+
+  module ClassMethods
+
+    # Resolve the custom field set matching our model type defined at the closest ancestor level.
+    # Note, this is overridden for CustomValueSet which has it's field set explicitly assigned based on the link context
+    def resolve_custom_field_set(division: nil, model: nil, required: true)
+      CustomFieldSet.resolve(self.name, division: division, model: model, required: required)
+    end
+
+
+    # Filter against a custom value
+    # Optionally specify a base relation to chain from.
+    # future: expand our filter API as needs are more clear
+    def where_custom_value(field_identifier, value, base_relation: nil, division: nil)
+      base_relation ||= self
+      field = resolve_custom_field_set(division: division).field(field_identifier)
+      base_relation.where("custom_data->>? = ?", field.json_key, value)
+    end
+
+    def custom_field?(field_identifier, division: nil, model: nil)
+      field_set = resolve_custom_field_set(division: division, model: model, required: false)
+      field_set && field_set.field(field_identifier, required: false).present?
+    end
+
+  end
+
+
+  def ensured_custom_data
+    self.custom_data ||= {}
+    self.custom_data  # this explicit return value is important!
+  end
+
+  def update_custom_value(field_identifier, value)
+    data = set_custom_value(field_identifier, value)
+    self.update(custom_data: data)
+  end
+
+  def set_custom_value(field_identifier, value)
+    field = custom_field(field_identifier)  # note, this is fatal if field not found
+    if field.translatable?
+      set_translation(field.json_key, value)
+    else
+      data = ensured_custom_data
+      # future: value manipulation depending on field data type
+      data[field.json_key] = value
+      data
+    end
+  end
+
+
+  def custom_value(field_identifier)
+    field = custom_field(field_identifier)
+    result = ensured_custom_data[field.json_key]
+    # future: result manipulation depending on field data type
+  end
+
+  # todo: consider making 'required: false' the default
+  def resolve_custom_field_set(required: true)
+    self.class.resolve_custom_field_set(model: self, required: required)
+  end
+
+  def custom_field(field_identifier, required: true)
+    field_set = resolve_custom_field_set(required: required)
+    field_set.field(field_identifier, required: required)  if field_set
+  end
+
+  def custom_field?(field_identifier)
+    self.class.custom_field?(field_identifier, model: self)
+  end
+
+  def method_missing(method_sym, *arguments, &block)
```

Comment:
Would it be possible to document these metaprogramming methods? There's quite a lot of complexity here.